node/network/collator-protocol/src/collator_side/mod.rs

```diff
@@ -569,6 +619,50 @@ where
 	.await;
 }
 
+/// Discover and connect to a set of validators in advance. Should only be issued
+/// if the collator is aware of its assignment on the child of the given relay parent.\
+/// Connects to both current and the next group if the block is
+/// [close](NEXT_GROUP_PRECONNECT_WINDOW) to the group rotation boundary.
+async fn preconnect<Context>(
+	ctx: &mut Context,
+	runtime: &mut RuntimeInfo,
+	para_id: ParaId,
+	relay_parent: Hash,
+) -> Result<()>
+where
+	Context: SubsystemContext<Message = CollatorProtocolMessage>,
+	Context: overseer::SubsystemContext<Message = CollatorProtocolMessage>,
+{
+	let (our_core, num_cores) =
+		match determine_core(ctx, para_id, relay_parent, AssignmentType::Lookahead).await? {
+			Some(core) => core,
+			None => {
+				gum::warn!(
+					target: LOG_TARGET,
+					para_id = %para_id,
+					?relay_parent,
+					"Preconnect failed: no occupied core found for the para"
+				);
+
+				return Ok(())
+			},
+		};
+
+	let validators_group = determine_our_validators(
+		ctx,
+		runtime,
+		our_core,
+		num_cores,
+		relay_parent,
+		AssignmentType::Lookahead,
+	)
+	.await?;
+
+	connect_to_validators(ctx, validators_group.validators.into_iter().collect()).await;
```

Comment:
This stateless connection mechanism can't really work, but the issue was present before already and could be an explanation for short stalls:

If we have two forks with different height, they might be in different backing groups, hence they would disconnect each other. I believe we had this discussion before ... @ordian I believe you had some reasoning, why this is not a problem?

Reply:
This might be a problem at group rotation boundaries. In #4640 (was reverted) we used a union of those groups. Maybe we should prefer the new group to the old one in this case?

Reply:
Doesn't making NEXT_GROUP_PRECONNECT_WINDOW big enough cover this issue? How big the spread of block number in our view can be?

Reply:
That would still lead to stalls, if the shorter fork is actually the better one. We should just make sure to be connected to both - I will have a look at that.

Reply:
> Doesn't making NEXT_GROUP_PRECONNECT_WINDOW big enough cover this issue? How big the spread of block number in our view can be?

it is not really limited, larger spreads are just becoming increasingly unlikely. Indeed for pre-connect it will likely be less of a problem, due to `NEXT_GROUP_PRECONNECT_WINDOW`.